### DISCUSSION TITLE
PKCS#7 vs CMS API

### DISCUSSION THREAD

**Discussion by beldmit** [openssl/openssl#25004] [General Discussion]
I'd like to discuss do we need to extend support for both PKCS#7 and CMS API or should we leave PKCS#7 as is. 

My personal view is that it's worth deprecating PKCS#7.

- PKCS#7 format is superseded by CMS format.
- There are no new specifications of PKCS#7.
- OpenSSL `cms` command can deal with the files created by `pkcs7` command
- Having 2 APIs for basically one purpose is bad.
- It can be easier to provide support of providers only in one API

**Comment by levitte** [comment]
I would be in favor of deprecating PKCS#7.  But then, do it for real, i.e. deprecate the symbols, deprecate `openssl pkcs7`, and document that they're likely not to work with more advanced providers (those who do need the parametrization that the old/legacy ctrls `EVP_PKEY_CTRL_PKCS7_{ENCRYPT,DECRYPT,SIGN}` provided).

Letting PKCS#7 simply *rot*, which is what we effectively would do otherwise, is not the answer.

**Reply by beldmit** [reply]
@levitte I have impression that it is already sort of rotten. Yes, formal deprecation is the best option

**Comment by tomato42** [comment]
+1 for just removing it ("let's nuke the entire site from orbit...")

**Comment by mattcaswell** [comment]
Deprecating it seems like a good idea. However that does have certain impacts. In particular:

- Our PKCS12 API uses the PCKS7 API, so we would have to implement CMS variants of those APIs
- Similarly the Timestamp protocol APIs (RFC3161) are also built on top of the PKCS7 API so we would need to introduce similar changes there

**Reply by levitte** [reply]
> * Similarly the Timestamp protocol APIs (RFC3161) are also built on top of the PKCS7 API so we would need to introduce similar changes there

Er..... no?  At least my read of [RFC3161](https://www.rfc-editor.org/rfc/rfc3161) shows me CMS, most notably in the ASN.1 module:

```
     ContentInfo FROM CryptographicMessageSyntax {iso(1)
     member-body(2) us(840) rsadsi(113549) pkcs(1) pkcs-9(9)
     smime(16) modules(0) cms(1)}
```

**Reply by levitte** [reply]
So if *OpenSSL*'s TS implementation is on top of PKCS#7, it's not quite right...  but not wrong either, we're just plain lucky that no CMS features not present in PKCS#7 are used.

**Reply by mattcaswell** [reply]
I said the *APIs* are based on it:

``` c
TS_TST_INFO *PKCS7_to_TS_TST_INFO(PKCS7 *token);
void TS_RESP_set_tst_info(TS_RESP *a, PKCS7 *p7, TS_TST_INFO *tst_info);
PKCS7 *TS_RESP_get_token(TS_RESP *a);
int TS_RESP_verify_signature(PKCS7 *token, STACK_OF(X509) *certs,
                             X509_STORE *store, X509 **signer_out);
int TS_RESP_verify_token(TS_VERIFY_CTX *ctx, PKCS7 *token);
```

My point is simply that we would have to replace the above functions with CMS based variants.

**Reply by levitte** [reply]
I understand.  You also referred to the RFC, which is what I'm commenting on.

And yeah, I arrive to the same conclusion, that an upgrade of the API is in order.  We really should, if nothing else to be compatible with the RFC!

**Reply by levitte** [reply]
PKCS#12 is a different story, unfortunately; [RFC7292](https://www.rfc-editor.org/rfc/rfc7292) refers quite firmly to PKCS#7 😞

**Reply by mattcaswell** [reply]
Are there any bad consequences to allowing CMS in PCKS#12?

**Reply by levitte** [reply]
I have no clue, really.  The bits taken from PKCS#7 seem to be ContentInfo and DigestInfo...  so it all depends on how those structures have "evolved".  But you know, "what's the harm" and "what could ever go wrong"? 😉
